### PR TITLE
support for p2p buffer transfers

### PR DESCRIPTION
**Pull request by wozeparrot**
need to wait till pyopencl tags a new release

adds a new buffer class: `RawBufferTransfer` that has a `transfer` method and a `_transfer` method that the subclass can override

gated behind envvar `P2P=1`

before:
```
0.01290 s, 2.60151 GB/s, 2048x2048x2, 33554.43200 KB
0.02536 s, 2.64595 GB/s, 2048x2048x4, 67108.86400 KB
0.04358 s, 3.07961 GB/s, 2048x2048x8, 134217.72800 KB
0.07394 s, 3.63060 GB/s, 2048x2048x16, 268435.45600 KB
0.14574 s, 3.68363 GB/s, 2048x2048x32, 536870.91200 KB
```
after:
```
0.00303 s, 11.05970 GB/s, 2048x2048x2, 33554.43200 KB
0.00457 s, 14.66897 GB/s, 2048x2048x4, 67108.86400 KB
0.00810 s, 16.56018 GB/s, 2048x2048x8, 134217.72800 KB
0.01551 s, 17.30784 GB/s, 2048x2048x16, 268435.45600 KB
0.02382 s, 22.53919 GB/s, 2048x2048x32, 536870.91200 KB
```

### CODE REVIEW

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/lazy.py                           5               2
tinygrad/runtime/lib.py                    9               0
tinygrad/runtime/ops_gpu.py                5               2
------------------------------------------------------------
total                                     19               4
------------------------------------------------------------
lines added in the tinygrad folder: 15
```

File: tinygrad/runtime/ops_gpu.py [hunk 1]
```diff
@@ -49,6 +49,9 @@ def _copyout(self, x:np.ndarray):
     buf = cl.Buffer(CL.cl_ctxs[self._buf.device], cl.mem_flags.WRITE_ONLY | cl.mem_flags.USE_HOST_PTR, 0, hostbuf=x.data)
     mapped, event = cl.enqueue_map_buffer(CL.cl_queue[self._buf.device], buf, cl.map_flags.WRITE, 0, self.size, dtype=self.dtype.np, is_blocking=False)
     with mapped.base: cl.enqueue_copy(CL.cl_queue[self._buf.device], mapped, self._buf, is_blocking=True, wait_for=[event])
+  def _transfer(self, x):
+    if "gfx" in CL.cl_ctxs[x._buf.device].devices[0].name: # TODO: only on amd
```
Comment (geohot) [inline tinygrad/runtime/ops_gpu.py:53]: Err, this will just silently do nothing on non AMD?

Comment (wozeparrot) [reply tinygrad/runtime/ops_gpu.py:53]: yea, currently it kinda does, thats why i gated it in lazy.py behind an envvar, im not sure what the best way to handle this is

**Comment by wozeparrot** [comment]
Release is now tagged upstream: https://github.com/inducer/pyopencl/pull/697#issuecomment-1678176387, just waiting on pypi wheels.

**Comment by geohot** [comment]
Nice work!

**Pull request merged**